File: ggcommons/ggcommons.py
"""
GGCommons main class with builder support.

This module provides the main GGCommons class. Component code accesses the
underlying subsystems through typed accessors (get_config_manager / get_messaging
/ get_metrics) rather than a service registry — matching the Java and Rust
libraries, which depend on the concrete ConfigManager / MessagingClient /
MetricEmitter directly.
"""

import argparse
import logging
from typing import Optional, List
from ggcommons.config.manager.config_manager import ConfigManager
from ggcommons.config.manager.config_manager_builder import ConfigManagerBuilder

logger = logging.getLogger(__name__)


class GGCommons:
    """
    Main entry point for the GGCommons framework with enhanced features.
    
    Provides dependency injection, service registry, and improved initialization.
    """

    def __init__(self, component_name: str, args: List[str], 
                 app_options: Optional[argparse.ArgumentParser] = None,
                 receive_own_messages: bool = True):
        """
        Initialize GGCommons with enhanced features.
        
        Args:
            component_name: The fully qualified component name
            args: Command line arguments
            app_options: Optional custom argument parser
            receive_own_messages: Whether to receive own messages (IPC only)
            
        Raises:
            ValueError: If component_name is None or empty
        """
        if not component_name:
            raise ValueError("Component name cannot be None or empty")
            
        self._component_name = component_name
        self._config_manager: Optional[ConfigManager] = None
        self._heartbeat = None

        try:
            # Process command line arguments
            parsed_args = self._process_args(component_name, args, app_options)

            # Initialize messaging FIRST: the GG_CONFIG / SHADOW / CONFIG_COMPONENT
            # config sources load the component configuration over messaging, so the
            # MessagingClient must be available before the config manager is built.
            self._init_messaging(parsed_args, receive_own_messages)

            # Initialize configuration manager
            self._init_config_manager(component_name, parsed_args)

            # Initialize metric emitter
            self._init_metrics()
            
            # Initialize heartbeat
            self._init_heartbeat()
            
            # Complete initialization
            if hasattr(self._config_manager, 'complete_initialization'):
                self._config_manager.complete_initialization()
                
            logger.info("GGCommons initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize GGCommons: {e}")
            # Tear down whatever was already started (messaging/metrics/heartbeat
            # threads, file watchers) so a failed init does not leak resources.
            # shutdown() is fully defensive, so it is safe on a partial init.
            try:
                self.shutdown()
            except Exception as cleanup_error:
                logger.error(f"Error during cleanup after failed init: {cleanup_error}")
            raise
            
    def _process_args(self, component_name: str, args: List[str], 
                     app_options: Optional[argparse.ArgumentParser]) -> argparse.Namespace:
        """
        Process command line arguments.
        
        Args:
            component_name: The component name
            args: Command line arguments
            app_options: Optional custom argument parser
            
        Returns:
            Parsed arguments namespace
        """
        parser = app_options or argparse.ArgumentParser()
        
        # Add standard ggcommons arguments
        parser.add_argument(
            '-c', '--config',
            nargs='*',
            type=str,
            default=['GG_CONFIG'],
            help='Configuration source. One of: ENV, GG_CONFIG, FILE, SHADOW, CONFIG_COMPONENT'
        )
        parser.add_argument(
            '-m', '--mode',
            nargs='*',
            type=str,
            help='Runtime mode - GREENGRASS (default) or STANDALONE <config_file_path>'
        )
        parser.add_argument(
            '-t', '--thing',
            type=str,
            help='Thing name to use (optional)'
        )
        
        parsed = parser.parse_args(args)
        
        # Process mode argument to match Java behavior
        if not hasattr(parsed, 'mode') or not parsed.mode:
            parsed.mode = ['GREENGRASS']

        mode_name = parsed.mode[0].upper()
        # Validate STANDALONE mode has config path
        if mode_name == 'STANDALONE':
            if len(parsed.mode) < 2:
                logger.error("STANDALONE mode requires config file path")
                raise ValueError("STANDALONE mode requires config file path")
        elif mode_name != 'GREENGRASS':
            # Reject unknown modes instead of silently treating them as GREENGRASS.
            logger.error(f"Unknown mode '{parsed.mode[0]}'")
            raise ValueError(
                f"Unknown mode '{parsed.mode[0]}'. Valid values are 'GREENGRASS' and 'STANDALONE'"
            )

        # Validate the config source token up front rather than failing later.
        valid_sources = {'FILE', 'ENV', 'GG_CONFIG', 'SHADOW', 'CONFIG_COMPONENT'}
        if parsed.config and parsed.config[0].upper() not in valid_sources:
            logger.error(f"Unrecognized config source '{parsed.config[0]}'")
            raise ValueError(
                f"Unrecognized config source '{parsed.config[0]}'. Valid values are "
                f"{', '.join(sorted(valid_sources))}"
            )

        return parsed
        
    def _init_config_manager(self, component_name: str, parsed_args: argparse.Namespace) -> None:
        """
        Initialize the configuration manager.
        
        Args:
            component_name: The component name
            parsed_args: Parsed command line arguments
        """
        # Use config manager builder to create appropriate manager
        self._config_manager = ConfigManagerBuilder.build(parsed_args, component_name)
        
    def _init_messaging(self, parsed_args: argparse.Namespace, receive_own_messages: bool) -> None:
        """
        Initialize the messaging client.
        
        Args:
            parsed_args: Parsed command line arguments
            receive_own_messages: Whether to receive own messages
        """
        # Import here to avoid circular imports
        from ggcommons.messaging.messaging_client import MessagingClient
        
        # Determine standalone config path
        standalone_config_path = None
        if hasattr(parsed_args, 'mode') and parsed_args.mode:
            if len(parsed_args.mode) > 1 and parsed_args.mode[0].upper() == 'STANDALONE':
                standalone_config_path = parsed_args.mode[1]
        
        MessagingClient.init(parsed_args, standalone_config_path, receive_own_messages)
        
    def _init_metrics(self) -> None:
        """Initialize the metric emitter."""
        # Import here to avoid circular imports
        from ggcommons.metrics.metric_emitter import MetricEmitter

        MetricEmitter.init(self._config_manager)

    def _init_heartbeat(self) -> None:
        """Initialize the heartbeat system, wiring it to the concrete subsystems."""
        # Import here to avoid circular imports
        from ggcommons.heartbeat.enhanced_heartbeat import EnhancedHeartbeat
        from ggcommons.messaging.messaging_client import MessagingClient
        from ggcommons.metrics.metric_emitter import MetricEmitter

        self._heartbeat = EnhancedHeartbeat(self._config_manager)
        # MessagingClient / MetricEmitter expose their operations as static methods,
        # so the classes themselves serve as the messaging/metric handles.
        self._heartbeat.set_messaging_service(MessagingClient)
        self._heartbeat.set_metric_service(MetricEmitter)
            
    def get_config_manager(self) -> ConfigManager:
        """
        Get the configuration manager instance.
        
        Returns:
            The configuration manager
            
        Raises:
            RuntimeError: If not properly initialized
        """
        if self._config_manager is None:
            raise RuntimeError("GGCommons not properly initialized")
        return self._config_manager
        
    def get_messaging(self):
        """
        Get the messaging handle (the MessagingClient class, whose operations are
        static). Mirrors Java's getMessaging() / Rust's messaging() accessor.

        Returns:
            The MessagingClient class
        """
        from ggcommons.messaging.messaging_client import MessagingClient
        return MessagingClient

    def get_metrics(self):
        """
        Get the metrics handle (the MetricEmitter class, whose operations are
        static). Mirrors Java's getMetrics() / Rust's metrics() accessor.

        Returns:
            The MetricEmitter class
        """
        from ggcommons.metrics.metric_emitter import MetricEmitter
        return MetricEmitter


    def shutdown(self) -> None:
        """
        Shutdown GGCommons and clean up resources.

        Each subsystem is closed independently so a failure in one does not leave
        the others leaking: heartbeat -> metrics -> messaging -> config (matching
        the Java shutdown order).
        """
        from ggcommons.messaging.messaging_client import MessagingClient
        from ggcommons.metrics.metric_emitter import MetricEmitter

        try:
            # Stop the heartbeat first so it stops publishing/emitting.
            if self._heartbeat and hasattr(self._heartbeat, 'stop'):
                self._heartbeat.stop()
        except Exception as e:
            logger.error(f"Error stopping heartbeat during shutdown: {e}")

        try:
            # Flush + stop the metric emitter's target thread.
            MetricEmitter.shutdown()
        except Exception as e:
            logger.error(f"Error shutting down metrics during shutdown: {e}")

        try:
            MessagingClient.shutdown()
        except Exception as e:
            logger.error(f"Error shutting down messaging during shutdown: {e}")

        try:
            # Stop the config manager's file-watcher thread (if any).
            if self._config_manager is not None:
                self._config_manager.close()
        except Exception as e:
            logger.error(f"Error closing config manager during shutdown: {e}")

        logger.info("GGCommons shutdown completed")

File: libs/ts/src/ggcommons.ts
/**
 * GGCommons (TypeScript) — library entry point / lifecycle.
 *
 * {@link GGCommonsBuilder} parses the standard CLI contract, initializes messaging
 * for the runtime mode, loads + validates configuration from the selected source,
 * initializes logging, metrics, and the heartbeat, and wires config hot-reload.
 * Mirrors the Rust `GgCommonsBuilder` / `GgCommons`.
 *
 * TypeScript has no RAII/Drop, so resources are released by {@link GGCommons.close}
 * (stops the heartbeat + config watch and disconnects messaging) rather than on GC.
 */
import { parseArgs, ParsedArgs } from "./cli";
import {
  Transport,
  profileLoggingFormat,
  profileHealthEnabled,
  profileMetricTarget,
  profileCredentialsKeyProvider,
} from "./platform";
import { Config } from "./config/model";
import { HealthServer, ReadinessState } from "./health";
import { resolve } from "./config/template";
import { validate } from "./config/validation";
import { buildConfigSource, ConfigSource, ConfigWatch } from "./config/source";
import { ConfigurationChangeListener } from "./config";
import { GgError } from "./errors";
import { Heartbeat } from "./heartbeat";
import { initLogging, reconfigureLogging, LoggingReconfigurer, logger } from "./logging";
import { DefaultMessagingService } from "./messaging/service";
import { IMessagingService } from "./messaging/types";
import { StandaloneMqttProvider } from "./messaging/standalone-provider";
import { IpcMessagingProvider } from "./messaging/ipc-provider";
import { loadMessagingConfig } from "./messaging/config";
import { MetricEmitter } from "./metrics/service";
import { MetricService } from "./metrics/types";
import type { StreamMetricsBridge, StreamService } from "./streaming";
import type { CredentialMetricsBridge, CredentialService } from "./credentials";
import type { ParameterService } from "./parameters";

/** The initialized component runtime: wired services + the current config snapshot. */
export class GGCommons {
  constructor(
    private readonly componentNameValue: string,
    private readonly argsValue: ParsedArgs,
    private current: Config,
    private readonly messagingService: IMessagingService | undefined,
    private readonly metricsService: MetricService,
    private readonly listeners: ConfigurationChangeListener[],
    private readonly heartbeat: Heartbeat,
    private readonly configSource: ConfigSource,
    private readonly readiness: ReadinessState,
  ) {}

  private configWatch?: ConfigWatch;
  private healthServer?: HealthServer;
  private signalHandlers: Array<[NodeJS.Signals, () => void]> = [];
  private closed = false;
  private streamsService?: StreamService;
  private streamMetrics?: StreamMetricsBridge;
  private credentialsService?: CredentialService;
  private credentialMetrics?: CredentialMetricsBridge;
  private parametersService?: ParameterService;

  /** @internal Attach the config-watch handle after construction. */
  _setWatch(watch: ConfigWatch | undefined): void {
    this.configWatch = watch;
  }

  /** @internal Attach the HTTP health server after construction (FR-HB-1). */
  _setHealth(server: HealthServer | undefined): void {
    this.healthServer = server;
  }

  /**
   * @internal Wire SIGTERM/SIGINT to graceful shutdown — the library owns this so a component never
   * leaks subscriptions on the shared connection (FR-HB-2). On signal: flip `/readyz` to 503
   * immediately, run the idempotent {@link close} path, then `process.exit(0)`. Handlers are removed in
   * {@link close} so a clean library shutdown leaves no listeners behind.
   */
  _installSignalHandlers(): void {
    for (const signal of ["SIGTERM", "SIGINT"] as NodeJS.Signals[]) {
      const handler = (): void => {
        this.readiness.beginShutdown();
        logger.info(`${signal} received; shutting down`);
        this.close()
          .catch((e) => logger.warn(`shutdown error: ${String(e)}`))
          .finally(() => process.exit(0));
      };
      process.on(signal, handler);
      this.signalHandlers.push([signal, handler]);
    }
  }

  /**
   * Set the app-controlled readiness flag consumed by `/readyz` and `/startupz` (FR-HB-1). Defaults to
   * `true`, so a component is ready as soon as messaging connects. Call `setReady(false)` early (before
   * subscribing to required topics) and `setReady(true)` once the component can serve, to gate traffic
   * on the component's own startup work. Mirrors the Rust/Java/Python `setReady`/`set_ready`.
   */
  setReady(ready: boolean): void {
    this.readiness.setReady(ready);
  }

  /** Whether the runtime is currently ready (`messaging connected && readyFlag && !shuttingDown`). */
  ready(): boolean {
    return this.readiness.isReady();
  }

  /** @internal Attach the streaming service + metrics bridge after construction. */
  _setStreaming(service: StreamService | undefined, bridge: StreamMetricsBridge | undefined): void {
    this.streamsService = service;
    this.streamMetrics = bridge;
  }

  /**
   * The telemetry streaming service, or `undefined` if the component config has no `streaming`
   * section. Obtain a stream with `service.stream(name)` and append durable records. Mirrors the
   * Rust/Java/Python `gg.streams()`.
   */
  streams(): StreamService | undefined {
    return this.streamsService;
  }

  /** @internal Attach the credential service after construction. */
  _setCredentials(service: CredentialService | undefined): void {
    this.credentialsService = service;
  }

  /** @internal Attach the credential metrics bridge after construction. */
  _setCredentialMetrics(bridge: CredentialMetricsBridge | undefined): void {
    this.credentialMetrics = bridge;
  }

  /**
   * The credential service, or `undefined` if the component config has no `credentials` section.
   * Mirrors the Rust `gg.credentials()` / Java/Python `getCredentials()`/`get_credentials()`.
   */
  credentials(): CredentialService | undefined {
    return this.credentialsService;
  }

  /** @internal Attach the parameter service after construction. */
  _setParameters(service: ParameterService | undefined): void {
    this.parametersService = service;
  }

  /**
   * The parameter service, or `undefined` if the component config has no `parameters` section.
   * Offline-first reads of externalized config (`get`/`getByPath`/typed accessors). Mirrors the
   * Rust `gg.parameters()`.
   */
  parameters(): ParameterService | undefined {
    return this.parametersService;
  }

  /** The component's full name. */
  componentName(): string {
    return this.componentNameValue;
  }

  /** The parsed standard CLI arguments. */
  args(): ParsedArgs {
    return this.argsValue;
  }

  /** A consistent snapshot of the current configuration (replaced atomically on reload). */
  config(): Config {
    return this.current;
  }

  /** The messaging service, or throw if none was wired. */
  messaging(): IMessagingService {
    if (!this.messagingService) {
      throw GgError.messaging("messaging is not available in this runtime mode");
    }
    return this.messagingService;
  }

  /** The metric service. */
  metrics(): MetricService {
    return this.metricsService;
  }

  /** Register a listener invoked after the configuration is hot-reloaded. */
  addConfigChangeListener(listener: ConfigurationChangeListener): void {
    this.listeners.push(listener);
  }

  /** Remove a previously-registered config-change listener (by identity). */
  removeConfigChangeListener(listener: ConfigurationChangeListener): void {
    const idx = this.listeners.indexOf(listener);
    if (idx >= 0) this.listeners.splice(idx, 1);
  }

  /**
   * Release resources: flip readiness to 503, unsubscribe/close every subsystem, and stop the health
   * server. Idempotent (safe to call from both the SIGTERM handler and the app). The shutting-down flag
   * is set first — even on a repeat call — so `/readyz` returns 503 the instant shutdown begins
   * (FR-HB-2), before the drain completes.
   */
  async close(): Promise<void> {
    this.readiness.beginShutdown();
    if (this.closed) return;
    this.closed = true;
    // Drop the library-owned SIGTERM/SIGINT handlers so a clean shutdown leaves no listeners.
    for (const [signal, handler] of this.signalHandlers) {
      process.removeListener(signal, handler);
    }
    this.signalHandlers = [];
    (this.parametersService as { close?: () => void } | undefined)?.close?.();
    this.credentialMetrics?.close();
    this.streamMetrics?.close();
    this.streamsService?.close();
    this.heartbeat.stop();
    if (this.configWatch) await this.configWatch.close().catch(() => undefined);
    await this.metricsService.shutdown().catch(() => undefined);
    if (this.messagingService instanceof DefaultMessagingService) {
      await this.messagingService.disconnect().catch(() => undefined);
    }
    // Stop the health server last so `/readyz` keeps serving 503 throughout the drain above.
    if (this.healthServer) await this.healthServer.stop().catch(() => undefined);
  }

  /** @internal Apply a reloaded snapshot and notify listeners. */
  _applyReload(snapshot: Config): void {
    this.current = snapshot;
  }
}

/** Fluent builder for {@link GGCommons} (the supported construction path). */
export class GGCommonsBuilder {
  private argv?: string[];
  private receiveOwn = true;

  constructor(private readonly componentNameValue: string) {}

  /**
   * Supply argv WITHOUT the node/script prefix (i.e. `process.argv.slice(2)`).
   * If not set, `process.argv.slice(2)` is used.
   */
  args(argv: string[]): this {
    this.argv = argv;
    return this;
  }

  /**
   * Whether the component receives messages it itself published (mirrors the
   * Java/Python/Rust `receiveOwnMessages` flag; default `true` =
   * RECEIVE_ALL_MESSAGES, the Java-canonical value — DESIGN-core §12 #2). Honored on the IPC
   * transport (the IPC ReceiveMode); on the MQTT transport the local broker delivers per its own
   * semantics.
   */
  receiveOwnMessages(value: boolean): this {
    this.receiveOwn = value;
    return this;
  }

  /** Parse args, load+validate config, init logging/messaging/metrics/heartbeat. */
  async build(): Promise<GGCommons> {
    const parsed = parseArgs(this.argv ?? process.argv.slice(2));
    // The resolver already applied the identity precedence (-t ▸ AWS_IOT_THING_NAME ▸ default).
    const thingName = parsed.thing;

    // Messaging is initialized first (it depends only on the resolved transport), and the
    // CONFIG_COMPONENT / GG_CONFIG / SHADOW sources need a handle to fetch config.
    const { service: messaging, ipcProvider } = await initMessaging(
      parsed.transport,
      parsed.messagingConfigPath,
      this.receiveOwn,
    );

    const source = buildConfigSource(parsed.config, {
      messaging,
      ipcProvider,
      thingName,
      componentName: this.componentNameValue,
    });

    const raw = await source.load();
    validate(raw);
    let current = Config.fromValue(this.componentNameValue, thingName, raw);

    // Thread the resolved platform's default logging format into the configurator (Phase 1c / FR-LOG-1):
    // a KUBERNETES pod with no `logging.ts_format` logs structured stdout-JSON, while explicit config
    // still wins and HOST/GREENGRASS keep today's console/text default. The platform is known here
    // (resolved at parse time) even though config loads after the resolver/messaging.
    initLogging(current, { formatDefault: profileLoggingFormat(parsed.platform) });
    logger.info(
      `GGCommons initialized: component=${this.componentNameValue} thing=${thingName} configSource=${source.sourceName()}`,
    );

    // Thread the resolved platform's default metric target into the metrics service (Phase 1c /
    // FR-MET-4): a KUBERNETES pod with no `metricEmission.target` selects the pull-based prometheus
    // target, while explicit config still wins and HOST/GREENGRASS keep the library default (`log`).
    // Same threading as the logging-format/health defaults — no resolver→ConfigManager dependency.
    const emitter = await MetricEmitter.create(current, messaging, profileMetricTarget(parsed.platform));
    const metrics: MetricService = emitter;

    const listeners: ConfigurationChangeListener[] = [emitter, new LoggingReconfigurer()];

    const heartbeat = Heartbeat.start(() => current, metrics, messaging);

    // Readiness state behind /readyz (FR-HB-1): messaging-connected (live getter; no wired service ⇒
    // not ready) AND the app's readyFlag (default true) AND not shutting down.
    const readiness = new ReadinessState(() => messaging?.connected() ?? false);

    // Build the runtime first so the reload closure can update its snapshot.
    let runtime: GGCommons;
    const onUpdate = (rawUpdate: unknown): void => {
      try {
        validate(rawUpdate);
      } catch (e) {
        logger.warn(`reloaded config failed validation; keeping previous: ${String(e)}`);
        return;
      }
      let next: Config;
      try {
        next = Config.fromValue(this.componentNameValue, thingName, rawUpdate);
      } catch (e) {
        logger.warn(`reloaded config could not be parsed; keeping previous: ${String(e)}`);
        return;
      }
      current = next;
      runtime._applyReload(next);
      reconfigureLogging(next);
      logger.info("configuration reloaded");
      for (const listener of [...listeners]) {
        // Guard both synchronous throws and rejected promises so one bad listener
        // can never break a hot reload (matches the other libraries).
        try {
          Promise.resolve(listener.onConfigurationChange(next)).catch((e) =>
            logger.warn(`config change listener threw: ${String(e)}`),
          );
        } catch (e) {
          logger.warn(`config change listener threw: ${String(e)}`);
        }
      }
    };

    runtime = new GGCommons(
      this.componentNameValue,
      parsed,
      current,
      messaging,
      metrics,
      listeners,
      heartbeat,
      source,
      readiness,
    );
    // Credentials / local vault (only when a `credentials` config section is present). Loaded
    // dynamically so components that don't use it pay nothing. Opened BEFORE streaming so the vault
    // is available to resolve `$secret` references in the streaming config (mirrors Rust lib.rs).
    let credentialService: CredentialService | undefined;
    let credentialsApi: typeof import("./credentials") | undefined;
    const credentialsRaw = current.raw["credentials"];
    if (credentialsRaw && typeof credentialsRaw === "object") {
      credentialsApi = await import("./credentials");
      const resolved = JSON.parse(resolve(current, JSON.stringify(credentialsRaw)));
      // Transparently namespace every key by <thingName>/<componentName> (collision-free).
      const namespace = `${current.thingName}/${this.componentNameValue}`;
      // Platform-default vault key-provider (FR-CRED-6, precedence FR-RT-3): when `keyProvider.type`
      // is absent, KUBERNETES defaults to the env/software-KEK custodian; an explicit type wins, and
      // this does NOT auto-enable credentials (we only reach here because a section is present).
      credentialService = await credentialsApi.openFromConfig(
        resolved,
        namespace,
        profileCredentialsKeyProvider(parsed.platform),
      );
      runtime._setCredentials(credentialService);
      const credentialMetrics = new credentialsApi.CredentialMetricsBridge(current, metrics, credentialService);
      runtime._setCredentialMetrics(credentialMetrics);
      logger.info("Credentials vault initialized");
    }

    // Parameters (only when a `parameters` config section is present). Independent, offline-first
    // service for externalized config — sibling of credentials. Loaded dynamically so components
    // that don't use it pay nothing. No namespacing of parameter keys (matches the Rust port; the
    // cache path is per-component templated below).
    const parametersRaw = current.raw["parameters"];
    if (parametersRaw && typeof parametersRaw === "object") {
      const parametersApi = await import("./parameters");
      const resolved = JSON.parse(resolve(current, JSON.stringify(parametersRaw)));
      const parameterService = await parametersApi.openFromConfig(resolved);
      runtime._setParameters(parameterService);
      logger.info("Parameters service initialized");
    }

    // Telemetry streaming (only when a `streaming` config section is present, so components that
    // don't use it never load the native addon). Loaded dynamically for the same reason.
    const streamingRaw = current.raw["streaming"];
    if (streamingRaw && typeof streamingRaw === "object") {
      const streaming = await import("./streaming");
      // Resolve `$secret` references against the vault before streaming consumes its config, so
      // secrets never land in the templated/logged config snapshot (mirrors Rust §7).
      let streamingValue = JSON.parse(resolve(current, JSON.stringify(streamingRaw)));
      if (credentialService && credentialsApi) {
        streamingValue = credentialsApi.resolveSecretRefs(streamingValue, credentialService);
      }
      const streamingJson = JSON.stringify(streamingValue);
      const svc = streaming.StreamService.open(streamingJson);
      const names = streaming.StreamService.streamNames(streamingJson);
      const bridge = names.length
        ? new streaming.StreamMetricsBridge(current, metrics, svc, names)
        : undefined;
      runtime._setStreaming(svc, bridge);
      logger.info(`Telemetry streaming initialized with ${names.length} stream(s)`);
    }

    // HTTP health endpoint (FR-HB-1). Precedence (FR-RT-3): explicit `health.enabled` ▸ platform
    // default (on for KUBERNETES, off for GREENGRASS/HOST). The platform is known here (resolved at
    // parse time), reusing the same threading as the logging default — no resolver→ConfigManager dep.
    const healthCfg = current.parsed.health;
    const healthEnabled = healthCfg.enabled ?? profileHealthEnabled(parsed.platform);
    if (healthEnabled) {
      // A health-endpoint problem must NEVER crash the component (health is auxiliary), and by this
      // point messaging/heartbeat/streams are already live — letting a bind failure reject build()
      // would also leak them. Log and continue without a health server, mirroring Java/Rust.
      try {
        const health = await HealthServer.start({
          port: healthCfg.port,
          paths: {
            liveness: healthCfg.livenessPath,
            readiness: healthCfg.readinessPath,
            startup: healthCfg.startupPath,
          },
          readiness,
        });
        runtime._setHealth(health);
      } catch (e) {
        logger.error(`health server failed to start (continuing without it): ${String(e)}`);
      }
    }

    // The library owns SIGTERM/SIGINT → graceful shutdown (FR-HB-2): flip /readyz to 503, run the
    // idempotent close() (unsubscribe all + bounded-close), then exit 0. Components no longer wire
    // their own handlers (the example skeleton's duplicate is removed).
    runtime._installSignalHandlers();

    // Attach the watch only after the runtime exists, so a reload that fires during
    // subscription setup has a valid runtime to update.
    runtime._setWatch(await source.watch(onUpdate));
    return runtime;
  }
}

/**
 * Initialize the messaging service + IPC provider handle for the resolved transport (DESIGN-core
 * §4.2 transport-injection site). Branches on the resolved {@link Transport}, not a legacy mode enum.
 */
async function initMessaging(
  transport: Transport,
  messagingConfigPath: string | undefined,
  receiveOwnMessages: boolean,
): Promise<{ service: IMessagingService | undefined; ipcProvider?: IpcMessagingProvider }> {
  if (transport === Transport.MQTT) {
    if (!messagingConfigPath) {
      throw GgError.messaging(
        "MQTT transport requires a messaging-config path (--transport MQTT <messaging_config.json>)",
      );
    }
    const mc = await loadMessagingConfig(messagingConfigPath);
    const provider = await StandaloneMqttProvider.connect(mc);
    return { service: new DefaultMessagingService(provider) };
  }
  // IPC (GREENGRASS)
  const provider = await IpcMessagingProvider.connect({ receiveOwnMessages });
  return { service: new DefaultMessagingService(provider), ipcProvider: provider };
}
